libjwt/jwt-gnutls.c

```diff
@@ -0,0 +1,231 @@
+
+#include <gnutls/gnutls.h>
+#include <gnutls/crypto.h>
+#include <gnutls/x509.h>
+#include <gnutls/abstract.h>
+
+#include <jansson.h>
+
+#include <jwt.h>
+
+#include "jwt-private.h"
+#include "b64.h"
+#include "config.h"
+
+/**
+ * libjwt encryption/decryption function definitions
+ */
+int jwt_sign_sha_hmac(jwt_t *jwt, char **out, unsigned int *len, const char *str) {
+  int alg;
+  
+  switch (jwt->alg) {
+    case JWT_ALG_HS256:
+      alg = GNUTLS_DIG_SHA256;
+      break;
+    case JWT_ALG_HS384:
+      alg = GNUTLS_DIG_SHA384;
+      break;
+    case JWT_ALG_HS512:
+      alg = GNUTLS_DIG_SHA512;
+      break;
+    default:
+      alg = GNUTLS_DIG_NULL;
+      break;
+  }
+  
+  if (alg != GNUTLS_DIG_NULL) {
+    *len = gnutls_hmac_get_len(alg);
+    (*out) = malloc(*len);
+    if (*out != NULL) {
+      if (!gnutls_hmac_fast(alg, jwt->key, jwt->key_len, str, strlen(str), (*out))) {
+        return 0;
+      } else {
+        return EINVAL;
+      }
+    } else {
+      return ENOMEM;
+    }
+  } else {
+    return EINVAL;
+  }
+}
+
+int jwt_verify_sha_hmac(jwt_t *jwt, const char *head, const char *sig) {
+  char * sig_check, * buf = NULL;
+  unsigned int len;
+  int res, buf_len;
+  base64_encodestate state;
+  
+  if (!jwt_sign_sha_hmac(jwt, &sig_check, &len, head)) {
+    buf = alloca(len * 2);
+    jwt_base64_init_encodestate(&state);
+    buf_len = jwt_base64_encode_block(sig_check, len, buf, &state);
+    buf_len += jwt_base64_encode_blockend(buf + buf_len, &state);
+    buf[buf_len] = '\0';
+
+    jwt_base64uri_encode(buf);
+    res = !strcmp(sig, buf)?0:EINVAL;
+    free(sig_check);
+  } else {
+    res = EINVAL;
+  }
+  return res;
+}
+
+int jwt_sign_sha_pem(jwt_t *jwt, char **out, unsigned int *len, const char *str) {
+  gnutls_x509_privkey_t key;
+  gnutls_privkey_t privkey;
+  gnutls_datum_t key_dat = {(void *) jwt->key, jwt->key_len}, body_dat = {(void*) str, strlen(str)}, sig_dat;
+  int res, pk_alg;
+  int alg;
+  gnutls_x509_crt_fmt_t format;
+  
+  switch (jwt->alg) {
+    case JWT_ALG_RS256:
+      alg = GNUTLS_DIG_SHA256;
+      pk_alg = GNUTLS_PK_RSA;
+      format = GNUTLS_X509_FMT_PEM;
+      break;
+    case JWT_ALG_RS384:
+      alg = GNUTLS_DIG_SHA384;
+      pk_alg = GNUTLS_PK_RSA;
+      format = GNUTLS_X509_FMT_PEM;
+      break;
+    case JWT_ALG_RS512:
+      alg = GNUTLS_DIG_SHA512;
+      pk_alg = GNUTLS_PK_RSA;
+      format = GNUTLS_X509_FMT_PEM;
+      break;
+    case JWT_ALG_ES256:
+      alg = GNUTLS_DIG_SHA256;
+      pk_alg = GNUTLS_PK_EC;
+      format = GNUTLS_X509_FMT_PEM;
+      break;
+    case JWT_ALG_ES384:
+      alg = GNUTLS_DIG_SHA384;
+      pk_alg = GNUTLS_PK_EC;
+      format = GNUTLS_X509_FMT_PEM;
+      break;
+    case JWT_ALG_ES512:
+      alg = GNUTLS_DIG_SHA512;
+      pk_alg = GNUTLS_PK_EC;
+      format = GNUTLS_X509_FMT_PEM;
+      break;
+    default:
+      alg = GNUTLS_DIG_NULL;
+      pk_alg = GNUTLS_PK_UNKNOWN;
+      format = GNUTLS_X509_FMT_PEM;
+      break;
+  }
+  
```

Comment:
Can you please break out this deeply nested if to single if's. This is too hard to follow and doesn't really serve a purpose to have it nested like this.